File: entree-pos/PrintingManager.m


#import "PrintingManager.h"
#import "entree_pos-Swift.h"

#import <StarPrinting/StarPrinting.h>

#define kReceiptPrinterMACAddressKey @"receipt_printer_mac_address"

@interface PrintingManager ()

@property NSMutableDictionary *printerCache;

@property NSDateFormatter *dateFormatter;
@property NSNumberFormatter *numberFormatter;

- (void)printerWithMACAddress:(NSString *)address completion:(void (^)(Printer *printer))completion;
- (void)sendPrintData:(PrintData *)printData toPrinterWithMACAddress:(NSString *)address;

@end

@implementation PrintingManager

- (instancetype)init {
    self = [super init];
    if (self) {
        _printerCache = [NSMutableDictionary dictionary];
        
        _dateFormatter = [[NSDateFormatter alloc] init];
        _dateFormatter.dateStyle = NSDateFormatterShortStyle;
        _dateFormatter.timeStyle = NSDateFormatterShortStyle;
        
        _numberFormatter = [[NSNumberFormatter alloc] init];
        _numberFormatter.numberStyle = NSNumberFormatterCurrencyStyle;
    }
    return self;
}

#pragma mark - Shared Manager

+ (instancetype)sharedManager {
    static PrintingManager *sharedManager = nil;
    @synchronized(self) {
        if (sharedManager == nil) {
            sharedManager = [[self alloc] init];
        }
    }
    return sharedManager;
}

#pragma mark - Receipt Printer MAC Address

+ (NSString *)receiptPrinterMACAddress {
    return [[NSUserDefaults standardUserDefaults] objectForKey:kReceiptPrinterMACAddressKey];
}

+ (void)setReceiptPrinterMACAddress:(NSString *)address {
    [[NSUserDefaults standardUserDefaults] setObject:address forKey:kReceiptPrinterMACAddressKey];
}

- (void)presentAlertControllerForPrintErrorWithMessage:(NSString *)message fromViewController:(UIViewController *)viewController {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Printing Error" message:message preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil]];
    [viewController presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Printer Communication

- (void)printerWithMACAddress:(NSString *)address completion:(void (^)(Printer *printer))completion {
    Printer *cachedPrinter = self.printerCache[address];
    if (cachedPrinter) {
        completion(cachedPrinter);
    }
    
    [Printer search:^(NSArray *found) {
        for (Printer *printer in found) {
            self.printerCache[printer.macAddress] = printer;
            
            if (printer.macAddress == address) {
                completion(printer);
            }
        }
    }];
}

- (void)sendPrintData:(PrintData *)printData toPrinterWithMACAddress:(NSString *)address {
    [self printerWithMACAddress:address completion:^(Printer *printer) {
        if (printer.status == PrinterStatusConnected) {
            [printer print:printData];
        } else {
            [printer connect:^(BOOL success) {
                if (success) {
                    [printer print:printData];
                } else{
                    NSString *message = [NSString stringWithFormat:@"Failed to connect to printer with MAC address %@.", address];
                    [self presentAlertControllerForPrintErrorWithMessage:message fromViewController:nil];
                }
            }];
        }
    }];
}

#pragma mark - Send To Kitchen

- (void)printPrintJobsForOrderItems:(NSArray *)orderItems party:(Party *)party server:(Employee *)server toGo:(BOOL)toGo fromViewController:(UIViewController *)viewController {
    // This is wildly complex and should be simplified if at all possible
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:@"Kitchen printer not found on network." preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil]];
    [viewController presentViewController:alertController animated:true completion:nil];
    
    NSMutableArray *printJobs = [NSMutableArray array];
    NSMutableDictionary *printJobObjectIDToOrderItemMap = [NSMutableDictionary dictionary];
    NSMutableDictionary *printerMACAddressToPrintJobsMap = [NSMutableDictionary dictionary];
    
    for (OrderItem *orderItem in orderItems) {
        [printJobs addObjectsFromArray:orderItem.menuItem.printJobs];
        for (PrintJob *printJob in orderItem.menuItem.printJobs) {
            printJobObjectIDToOrderItemMap[printJob.objectId] = orderItem;
            printerMACAddressToPrintJobsMap[printJob.printer.mac] = [NSMutableArray array];
        }
    }
    
    for (PrintJob *printJob in printJobs) {
        NSMutableArray *printJobsForPrinter = printerMACAddressToPrintJobsMap[printJob.printer.mac];
        [printJobsForPrinter addObject:printJob];
    }
    
    for (NSString *address in printerMACAddressToPrintJobsMap.allKeys) {
        NSMutableDictionary *dictionary = [NSMutableDictionary dictionaryWithDictionary:@{@"{{location}}": toGo ? @"To Go" : @"Dine-In",
                                                                                          @"{{time}}": [self.dateFormatter stringFromDate:[NSDate date]],
                                                                                          @"{{table}}": party ? party.table.name : @"N/A",
                                                                                          @"{{server}}": party ? party.server.name : server.name}];
        
        NSMutableString *menuItemsXML = [NSMutableString string];
        
        NSArray *printJobsForPrinter = printerMACAddressToPrintJobsMap[address];
        for (PrintJob *printJob in printJobsForPrinter) {
            OrderItem *orderItem = printJobObjectIDToOrderItemMap[printJob.objectId];
            
            NSString *itemNameString = printJob.text.length == 0 ? printJob.text : orderItem.menuItem.name;
            NSString *itemNameXML = [NSString stringWithFormat:@"<text><left><large>%@</large></left></text><newline />", itemNameString];
            
            NSString *seatXML = orderItem.seatNumber == 0 ? @"" : [NSString stringWithFormat:@"<tab /><text><large>SEAT: %ld</large></text><newline />", (long) orderItem.seatNumber];
            
            NSMutableString *modifiersXML = [NSMutableString string];
            if (orderItem.menuItemModifiers.count > 0) {
                for (MenuItemModifier *modifier in orderItem.menuItemModifiers) {
                    NSString *modifierXML = [NSString stringWithFormat:@"<tab /><text><large><ic>+ %@</ic></large></text><newline />", modifier.name];
                    [modifiersXML appendString:modifierXML];
                }
            }
            
            NSString *notesXML = orderItem.notes.length == 0 ? @"" : [NSString stringWithFormat:@"<tab /><text><large><ic>+ %@</ic></large></text><newline />", orderItem.notes];
            
            NSString *menuItemXML = [NSString stringWithFormat:@"%@%@%@%@", itemNameXML, seatXML, modifiersXML, notesXML];
            [menuItemsXML appendString:menuItemXML];
        }
        
        dictionary[@"{{menu_items}}"] = menuItemsXML;
        
        NSString *filePath = [[NSBundle mainBundle] pathForResource:@"order_item_template" ofType:@"xml"];
        PrintData *printData = [[PrintData alloc] initWithDictionary:dictionary atFilePath:filePath];
        [self sendPrintData:printData toPrinterWithMACAddress:address];
    }
}

#pragma mark - Print Check

- (void)printCheckForOrderItems:(NSArray *)orderItems party:(Party *)party fromViewController:(UIViewController *)viewController {
    NSString *address = [PrintingManager receiptPrinterMACAddress];
    if (!address) {
        [self presentAlertControllerForPrintErrorWithMessage:@"No receipt printer set. Please configure this in settings." fromViewController:viewController];
        return;
    }
    
    Restaurant *restaurant = [Restaurant defaultRestaurantFromLocalDatastoreFetchIfNil];
    
    NSMutableString *orderItemsXML = [NSMutableString string];
    
    double amountDue = 0;
    double tax = 0;
    
    for (OrderItem *item in orderItems) {
        NSString *price = [self.numberFormatter stringFromNumber:[NSNumber numberWithDouble:[item itemCost]]];
        NSString *itemXML = [NSString stringWithFormat:@"<left><text>[%@] %@</text></left><newline />", price, item.menuItem.name];
        [orderItemsXML appendString:itemXML];
        
        amountDue += [item itemCost];
        tax += [item applicableTax];
    }

    double subtotal = amountDue + tax;
    
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionaryWithDictionary:@{@"{{restaurant}}": restaurant.name,
                                                                                      @"{{location}}": restaurant.location,
                                                                                      @"{{phone}}": restaurant.phone,
                                                                                      @"{{time}}": [self.dateFormatter stringFromDate:[NSDate date]],
                                                                                      @"{{table}}": party.table.name,
                                                                                      @"{{server}}": party.server.name,
                                                                                      @"{{items}}": orderItemsXML,
                                                                                      @"{{amount_due}}": [self.numberFormatter stringFromNumber:@(amountDue)],
                                                                                      @"{{tax}}": [self.numberFormatter stringFromNumber:@(tax)],
                                                                                      @"{{subtotal}}": [self.numberFormatter stringFromNumber:@(subtotal)]}];
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"check_template" ofType:@"xml"];
    PrintData *printData = [[PrintData alloc] initWithDictionary:dictionary atFilePath:filePath];
    [self sendPrintData:printData toPrinterWithMACAddress:address];
}

#pragma mark - Print Receipt

- (void)printReceiptForOrder:(Order *)order fromViewController:(UIViewController *)viewController {
    NSString *address = [PrintingManager receiptPrinterMACAddress];
    if (!address) {
        [self presentAlertControllerForPrintErrorWithMessage:@"No receipt printer set. Please configure this in settings." fromViewController:viewController];
        return;
    }
    
    BOOL cardPayment = order.payment.cardFlightChargeToken == nil;
    
    Restaurant *restaurant = [Restaurant defaultRestaurantFromLocalDatastoreFetchIfNil];
    
    NSMutableString *orderItemsXML = [NSMutableString string];
    for (OrderItem *orderItem in order.orderItems) {
        NSString *orderItemXML = [NSString stringWithFormat:@"<text>[%@] %@</text><newline />", [self.numberFormatter stringFromNumber:@(orderItem.itemCost)], orderItem.menuItem.name];
        [orderItemsXML appendString:orderItemXML];
    }
    
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionaryWithDictionary:@{@"{{copy}}": @"Customer Copy",
                                                                                      @"{{restaurant}}": restaurant.name,
                                                                                      @"{{location}}": restaurant.location,
                                                                                      @"{{phone}}": restaurant.phone,
                                                                                      @"{{time}}": [self.dateFormatter stringFromDate:[NSDate date]],
                                                                                      @"{{server}}": order.server.name,
                                                                                      @"{{order_id}}": order.objectId,
                                                                                      @"{{payment_id}}": order.payment.objectId,
                                                                                      @"{{order_items}}": orderItemsXML,
                                                                                      @"{{subtotal}}": [self.numberFormatter stringFromNumber:@(order.subtotal)]}];
    
    if (cardPayment) {
        dictionary[@"{{last_four}}"] = order.payment.cardLastFour;
        dictionary[@"{{cardholder}}"] = order.payment.cardName;
    } else {
        dictionary[@"{{cash_amount_paid}}"] = [self.numberFormatter stringFromNumber:[NSNumber numberWithDouble:order.payment.cashAmountPaid]];
        dictionary[@"{{change_given}}"] = [self.numberFormatter stringFromNumber:[NSNumber numberWithDouble:order.payment.changeGiven]];
    }
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:[NSString stringWithFormat:@"%@_receipt_template", order.payment.type.lowercaseString] ofType:@"xml"];
    PrintData *printData = [[PrintData alloc] initWithDictionary:dictionary atFilePath:filePath];
    [self sendPrintData:printData toPrinterWithMACAddress:address];
    
    if (cardPayment) {
        dictionary[@"{{copy}}"] = @"Merchant Copy";
        PrintData *printData = [[PrintData alloc] initWithDictionary:dictionary atFilePath:filePath];
        [self sendPrintData:printData toPrinterWithMACAddress:address];
    }
}

#pragma mark - Print Server Summary

- (void)printSummaryForServer:(Employee *)server date:(NSDate *)date fromViewController:(UIViewController *)viewController {
    NSString *address = [PrintingManager receiptPrinterMACAddress];
    if (!address) {
        [self presentAlertControllerForPrintErrorWithMessage:@"No receipt printer set. Please configure this in settings." fromViewController:viewController];
        return;
    }
    
    PFQuery *query = [Order query];
    query.limit = 1000;
    
    [query includeKey:@"payment"];
    
    [query whereKey:@"createdAt" greaterThanOrEqualTo:[[NSCalendar currentCalendar] startOfDayForDate:date]];
    [query whereKey:@"createdAt" lessThan:[[NSCalendar currentCalendar] startOfDayForDate:[date dateByAddingTimeInterval:86400]]];
    [query whereKey:@"server" equalTo:server];
    
    [query whereKeyExists:@"payment"];
    
    /*
    PFQuery *innerQuery = [Payment query];
    [innerQuery whereKey:@"chargedAt" equalTo:@YES];
    [query whereKey:@"payment" matchesQuery:innerQuery];
     */
    
    NSArray *orders = [query findObjects]; // This is a synchronous request, should be avoided if possible
    
    NSMutableString *ordersXML = [NSMutableString string];
    
    double cardTotal = 0;
    double cashTotal = 0;
    double tips = 0;
    
    for (Order *order in orders) {
        NSString *orderID = order.objectId;
        NSString *paymentType = order.payment.type;
        NSString *subtotal = [self.numberFormatter stringFromNumber:@(order.subtotal)];
        NSString *tip = [self.numberFormatter stringFromNumber:@(order.tip)];
        
        if ([paymentType isEqualToString:@"Card"]) {
            cardTotal += order.total;
        } else {
            cashTotal += order.total;
        }
        tips += order.tip;
        
        NSString *orderXML = [NSString stringWithFormat:@"<text>Order ID: %@</text><newline /><tab><text>Payment Type: %@</text><newline /><tab><text>Subtotal: %@</text><newline /><tab><text>Tip: %@</text><newline /><newline />", orderID, paymentType, subtotal, tip];
        [ordersXML appendString:orderXML];
    }
    
    NSDictionary *dictionary = @{@"{{name}}": server.name,
                                 @"{{date}}": [self.dateFormatter stringFromDate:date],
                                 @"{{orders}}": ordersXML,
                                 @"{{card}}": [self.numberFormatter stringFromNumber:@(cardTotal)],
                                 @"{{cash}}": [self.numberFormatter stringFromNumber:@(cashTotal)],
                                 @"{{tips}}": [self.numberFormatter stringFromNumber:@(tips)]};
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"server_summary_template" ofType:@"xml"];
    PrintData *printData = [[PrintData alloc] initWithDictionary:dictionary atFilePath:filePath];
    [self sendPrintData:printData toPrinterWithMACAddress:address];
}

#pragma mark - Printer Search Wrapper

- (void)searchForPrintersWithCompletion:(void (^)(NSArray<Printer *> *results))completion {
    [Printer search:^(NSArray *found) {
        completion(found);
    }];
}

@end

File: entree-pos/PrintingManager.swift

import UIKit
import StarKit

enum PrintingManagerError: ErrorType {
    case NoReceiptPrinter
    case PrinterNotFound
}

public class PrintingManager {
    
    public static var receiptPrinterMACAddress: String? {
        get {
            return NSUserDefaults.standardUserDefaults().objectForKey("receipt_printer_mac_address") as? String
        }
        set {
            NSUserDefaults.standardUserDefaults().setObject(newValue, forKey: "receipt_printer_mac_address")
        }
    }
    
    private static var printerCache = [String: Printer]()
    
    private static var dateFormatter: NSDateFormatter {
        let formatter = NSDateFormatter()
        formatter.dateStyle = .ShortStyle
        formatter.timeStyle = .ShortStyle
        return formatter
    }
    
    private static var currencyNumberFormatter: NSNumberFormatter {
        let formatter = NSNumberFormatter()
        formatter.numberStyle = .CurrencyStyle
        return formatter
    }
    
    // MARK: - PrintingManager
    
    private static func getPrinterWithMACAddress(address: String, handler: (Printer?) -> ()) {
        if let cachedPrinter = PrintingManager.printerCache[address] {
            handler(cachedPrinter)
            return
        }
        
        Printer.searchForLANPrintersWithCompletionHandler { printers in
            for printer in printers {
                PrintingManager.printerCache[printer.macAddress] = printer
            }
            
            if let printer = PrintingManager.printerCache[address] {
                handler(printer)
            } else {
                handler(nil)
            }
        }
    }
    
    private static func sendPrintData(printData: PrintData, toPrinterWithMACAddress address: String) {
        getPrinterWithMACAddress(address) { printer in
            printer?.printData(printData, completionHandler: {})
        }
    }
    
    // MARK: - Receipt printing
    
    public static func printReceiptForOrder(order: Order) throws {
        guard let address = PrintingManager.receiptPrinterMACAddress else {
            throw PrintingManagerError.NoReceiptPrinter
        }
        
        let isCardPayment = order.payment?.type == "Card"
        
        // Synchronously fetch restaurant for header information
        let defaultRestuarant = Restaurant.defaultRestaurantFromLocalDatastoreFetchIfNil()!
        
        var orderItemsXML = ""
        for orderItem in order.orderItems {
            let itemCost = currencyNumberFormatter.stringFromDouble(orderItem.itemCost())!
            orderItemsXML += "[\(itemCost)] \(orderItem.menuItem.name)<newline />"
        }
        
        var dictionary = [
            "{{copy}}": "Customer Copy",
            "{{restaurant}}": defaultRestuarant.name,
            "{{location}}": defaultRestuarant.location,
            "{{phone}}": defaultRestuarant.phone,
            "{{time}}": dateFormatter.stringFromDate(NSDate()),
            "{{server}}": order.server.name,
            "{{order_id}}": order.objectId!,
            "{{payment_id}}": order.payment!.objectId!,
            "{{order_items}}": orderItemsXML,
            "{{subtotal}}": currencyNumberFormatter.stringFromDouble(order.subtotal)!
        ]
        
        if isCardPayment {
            dictionary["{{last_four}}"] = order.payment!.cardLastFour!
            dictionary["{{cardholder}}"] = order.payment!.cardName!
        } else {
            dictionary["{{cash_amount_paid}}"] = currencyNumberFormatter.stringFromDouble(order.payment!.cashAmountPaid)
            dictionary["{{change_given}}"] = currencyNumberFormatter.stringFromDouble(order.payment!.changeGiven)
        }
        
        let filePath = NSBundle.mainBundle().pathForResource("\(order.payment!.type.lowercaseString)_receipt_template", ofType: "xml")!
        
        let printData = PrintData(dictionary: dictionary, filePath: filePath)
        
        PrintingManager.sendPrintData(printData, toPrinterWithMACAddress: address)
        
        if isCardPayment {
            dictionary["{{copy}}"] = "Merchant Copy"
            let printData = PrintData(dictionary: dictionary, filePath: filePath)
            PrintingManager.sendPrintData(printData, toPrinterWithMACAddress: address)
        }

    }
    
    // MARK: - Check printing
    
    public static func printCheckForOrderItems(orderItems: [OrderItem], party: Party) throws {
        guard let address = PrintingManager.receiptPrinterMACAddress else {
            throw PrintingManagerError.NoReceiptPrinter
        }
        
        let checkTemplateFilePath = NSBundle.mainBundle().pathForResource("check_template", ofType: "xml")!
        
        // Synchronously fetch restaurant for header information
        let defaultRestuarant = Restaurant.defaultRestaurantFromLocalDatastoreFetchIfNil()!
        
        // Stitch together template language for each order item
        var orderItemsXML = ""
        for orderItem in orderItems {
            
            let itemName = orderItem.menuItem.name
            let price = currencyNumberFormatter.stringFromDouble(orderItem.itemCost())!
            
            orderItemsXML += "<left><text>\(itemName)</text></left><newline /><right><text>\(price)</text></right><newline />"
        }
        
        // Calculate totals for the bottom of the check
        let amountDue = orderItems.reduce(0) {
            (amountDue, item) in
            return amountDue + item.itemCost()
        }
        let tax = orderItems.reduce(0) {
            (tax, item) in
            return tax + item.applicableTax()
        }
        let subtotal = amountDue + tax
        
        let dictionary = [
            "{{restaurant}}": defaultRestuarant.name,
            "{{location}}": defaultRestuarant.location,
            "{{phone}}": defaultRestuarant.phone,
            "{{time}}": dateFormatter.stringFromDate(NSDate()),
            "{{server}}": party.server.name,
            "{{items}}": orderItemsXML,
            "{{amount_due}}": currencyNumberFormatter.stringFromNumber(NSNumber(double: amountDue))!,
            "{{tax}}": currencyNumberFormatter.stringFromNumber(NSNumber(double: tax))!,
            "{{subtotal}}": currencyNumberFormatter.stringFromNumber(NSNumber(double: subtotal))!
        ]
        
        let printData = PrintData(dictionary: dictionary, filePath: checkTemplateFilePath)
        
        PrintingManager.sendPrintData(printData, toPrinterWithMACAddress: address)
    }
    
    // MARK: - Kitchen printing
    
    public static func printOrderItems(orderItems: [OrderItem], party: Party?, server: Employee, toGo: Bool) {
        // FIXME: Wildly complex and inefficient way to do this
        
        var printJobs = [PrintJob]()
        var printJobObjectIdToOrderItemMap = [String: OrderItem]()
        var printerMACAddressToPrintJobsMap = [String: [PrintJob]]()
        
        for orderItem in orderItems {
            printJobs += orderItem.menuItem.printJobs
            for printJob in orderItem.menuItem.printJobs {
                printJobObjectIdToOrderItemMap[printJob.objectId!] = orderItem
            }
        }
        
        for printJob in printJobs {
            if printerMACAddressToPrintJobsMap[printJob.printer.mac] != nil {
                printerMACAddressToPrintJobsMap[printJob.printer.mac]!.append(printJob)
            } else {
                printerMACAddressToPrintJobsMap[printJob.printer.mac] = [printJob]
            }
        }
        
        for (printerAddress, _) in printerMACAddressToPrintJobsMap {
            var dictionary = [
                "{{location}}": toGo ? "To Go" : "Dine-In",
                "{{time}}": dateFormatter.stringFromDate(NSDate()),
                "{{table}}": party?.table?.name ?? "",
                "{{server}}": server.name,
            ]
            
            var menuItemsTemplate = ""
            
            let printJobsForPrinter = printerMACAddressToPrintJobsMap[printerAddress]!

            for printJob in printJobsForPrinter {
                let orderItem = printJobObjectIdToOrderItemMap[printJob.objectId!]!
                
                let menuItemXML = "<left /><large>-" + (printJob.text.isEmpty ? orderItem.menuItem.name : printJob.text) + "</large><newline />"
                
                let seatXML = orderItem.seatNumber == 0 ? "" : "<tab /><large>+ SEAT: \(orderItem.seatNumber)</large><newline />"
                
                var modsXML = ""
                if !orderItem.menuItemModifiers.isEmpty {
                    modsXML = orderItem.menuItemModifiers.reduce("") {
                        (previous, modifier) in
                        return "\(previous)<tab /><large><invertcolor>+ \(modifier.name)</invertcolor></large><newline />"
                    }
                }
                
                let notesXML = orderItem.notes.isEmpty ? "" : "<tab /><large><invertcolor>+ \(orderItem.notes)</invertcolor></large><newline />"
                
                let menuItemTemplate = menuItemXML + seatXML + modsXML + notesXML
                menuItemsTemplate += menuItemTemplate
            }
            
            dictionary["{{menu_items}}"] = menuItemsTemplate
            
            let orderItemTemplateFilePath = NSBundle.mainBundle().pathForResource("order_item_template", ofType: "xml")!
            let printData = PrintData(dictionary: dictionary, filePath: orderItemTemplateFilePath)
            
            PrintingManager.sendPrintData(printData, toPrinterWithMACAddress: printerAddress)
        }
    }
    
    public static func printDailySummaryForServer(server: Employee, date: NSDate) throws {
        guard let address = PrintingManager.receiptPrinterMACAddress else {
            throw PrintingManagerError.NoReceiptPrinter
        }
        
        let query = Order.query()
        query?.limit = 1000
        
        query?.includeKey("payment")
        
        query?.whereKey("createdAt", greaterThanOrEqualTo: NSCalendar.currentCalendar().startOfDayForDate(date))
        query?.whereKey("createdAt", lessThan: NSCalendar.currentCalendar().startOfDayForDate(date.dateByAddingTimeInterval(86400)))
        query?.whereKey("server", equalTo: server)
        
        query?.whereKeyExists("payment")
        
        let orders = try query?.findObjects() as! [Order]
        
        let ordersTemplate = orders.reduce("") { previous, order in
            let orderID = order.objectId!
            let paymentType = order.payment!.type
            let subtotal = currencyNumberFormatter.stringFromDouble(order.subtotal)!
            let tip = currencyNumberFormatter.stringFromDouble(order.tip)!
            return "\(previous)Order ID: \(orderID)<newline /><tab>Payment Type: \(paymentType)<newline /><tab>Subtotal: \(subtotal)<newline /><tab>Tip: \(tip)<newline /><newline />"
        }
        
        let cardTotal = orders.reduce(0) { total, order in
            return total + (order.payment!.type == "Card" ? order.subtotal : 0)
        }
        let cashTotal = orders.reduce(0) { total, order in
            return total + (order.payment!.type == "Cash" ? order.subtotal : 0)
        }
        let tips = orders.reduce(0) { total, order in
            return total + order.tip
        }
        
        let dictionary = [
            "{{name}}": server.name,
            "{{date}}": dateFormatter.stringFromDate(date),
            "{{orders}}": ordersTemplate,
            "{{card}}": currencyNumberFormatter.stringFromDouble(cardTotal)!,
            "{{cash}}": currencyNumberFormatter.stringFromDouble(cashTotal)!,
            "{{tips}}": currencyNumberFormatter.stringFromDouble(tips)!
        ]
        
        let filePath = NSBundle.mainBundle().pathForResource("server_summary_template", ofType: "xml")!
        let printData = PrintData(dictionary: dictionary, filePath: filePath)
        
        PrintingManager.sendPrintData(printData, toPrinterWithMACAddress: address)
    }
    
}